communication/src/AsyncTlsConnection.cpp

```diff
@@ -0,0 +1,390 @@
+// Concord
+//
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+//
+// This product is licensed to you under the Apache 2.0 license (the "License").
+// You may not use this product except in compliance with the Apache 2.0 License.
+//
+// This product may include a number of subcomponents with separate copyright
+// notices and license terms. Your use of these subcomponents is subject to the
+// terms and conditions of the subcomponent's license, as noted in the
+// LICENSE file.
+
+#include <regex>
+
+#include <arpa/inet.h>
+#include <boost/filesystem.hpp>
+
+#include "AsyncTlsConnection.h"
+#include "TlsTcpImpl.h"
+
+namespace bft::communication {
+
+void AsyncTlsConnection::readMsgSizeHeader() {
+  auto self = shared_from_this();
+  async_read(*socket_, boost::asio::buffer(read_size_buf_), [this, self](const auto& error_code, auto _) {
+    if (error_code) {
+      if (error_code == boost::asio::error::operation_aborted || disposed_) {
+        // The socket has already been cleaned up and any references are invalid. Just return.
+        return;
+      } else {
+        // Remove the connection as it is no longer valid, and then close it, cancelling any ongoing operations.
+        LOG_ERROR(logger_,
+                  "Reading message size header failed for node " << peer_id_.value() << ": " << error_code.message());
+        return dispose();
+      }
+    }
+    // The message size header was read successfully.
+    if (getReadMsgSize() > tlsTcpImpl_.config_.bufferLength) {
+      LOG_ERROR(logger_,
+                "Message Size: " << getReadMsgSize() << " exceeds maximum: " << tlsTcpImpl_.config_.bufferLength
+                                 << " for node " << peer_id_.value());
+      return dispose();
+    }
+    readMsg();
+  });
+}
+
+void AsyncTlsConnection::readMsg() {
+  auto msg_size = getReadMsgSize();
+  read_msg_ = std::vector<char>(msg_size, 0);
+  auto self = shared_from_this();
+  startReadTimer();
+  async_read(*socket_, boost::asio::buffer(read_msg_), [this, self](const auto& error_code, auto _) {
+    if (error_code) {
+      if (error_code == boost::asio::error::operation_aborted || disposed_) {
+        // The socket has already been cleaned up and any references are invalid. Just return.
+        return;
+      } else {
+        // Remove the connection as it is no longer valid, and then close it, cancelling any ongoing operations.
+        LOG_ERROR(logger_,
+                  "Reading message of size <<" << getReadMsgSize() << " failed for node " << peer_id_.value() << ": "
+                                               << error_code.message());
+        return dispose();
+      }
+    }
+
+    // The Read succeeded.
+    boost::system::error_code _ec;
+    read_timer_.cancel(_ec);
```

Comment:
Do we want to make sure that the `_ec` does not return an error?
We might want to log it.

Reply:
I'm not sure it makes sense to log here.There's really nothing we can do, and it's not fatal. An error is only created when the handlers:

-   have already been invoked; or
-   have been queued for invocation in the near future.

While the latter is something we would never expect, the former may end up being a common case and littering our log with unnecessary warnings. I prefer to leave this as is.

Relatedly, I feel like `_` is a pretty common convention and prefer to leave that as well.